middleware/calc_metrics.go

```diff
@@ -0,0 +1,245 @@
+					"createdOn": now}}})
+
+	if err != nil {
+		logger.Log.WithFields(logrus.Fields{"msg": "Error updating metrics in record match run",
+			"rec match run ID": recMatchJob.ID,
+			"error":            err}).Warn("calcMetrics")
+		return err
+	}
+
+	return nil
+}
+
+func indexOf(s []string, e string) int {
+	for i, a := range s {
+		if a == e {
+			return i
+		}
+	}
+	return -1
+}
+
+func getAnswerKey(db *mgo.Database, recMatchJob *ptm_models.RecordMatchJob) (*fhir_models.Bundle, error) {
+	var answerKey *fhir_models.Bundle
+
+	// If deduplication mode, try to find an answer key w/ the master record set
+	if recMatchJob.MatchingMode == ptm_models.Deduplication {
+		masterRecSet := &ptm_models.RecordSet{}
+		c := db.C(ptm_models.GetCollectionName("RecordSet"))
+		// retrieve the master record set
+		err := c.Find(
+			bson.M{"_id": recMatchJob.MasterRecordSetID}).One(masterRecSet)
+		if err != nil {
+			logger.Log.WithFields(logrus.Fields{"func": "updateRecordMatchJob",
+				"err":          err,
+				"msg":          "Unable to find master record set",
+				"record setid": recMatchJob.MasterRecordSetID}).Warn("calcMetrics")
+			return nil, err
+		}
+
+		logger.Log.WithFields(logrus.Fields{
+			"rec set":            masterRecSet,
+			"answer key entries": len(masterRecSet.AnswerKey.Entry)}).Info("calcMetrics")
+
+		if len(masterRecSet.AnswerKey.Entry) > 1 {
+			answerKey = &masterRecSet.AnswerKey
+			logger.Log.WithFields(logrus.Fields{
+				"answer key entries": len(masterRecSet.AnswerKey.Entry)}).Info("calcMetrics")
+		}
+	} else if recMatchJob.MatchingMode == ptm_models.Query {
+		logger.Log.WithFields(logrus.Fields{
+			"msg": "Calculating Metrics for Query Mode Not Supoprted"}).Warn("calcMetrics")
+		return nil, nil
+	}
+
+	return answerKey, nil
+}
+
+func buildAnswerMap(answerKey *fhir_models.Bundle) (map[string]*groundTruth, int) {
+	totalResults := 0
+	m := make(map[string]*groundTruth)
+	for i, entry := range answerKey.Entry {
+		//		rtype := reflect.TypeOf(entry.Resource)
+		//		logger.Log.WithFields(logrus.Fields{
+		//			"answer":     i,
+		//			"entry type": rtype,
+		//			"entry kind": reflect.ValueOf(entry.Resource).Kind()}).Info("calcMetrics")
+
+		// Results are in untyped entry w/ links and search result
+		if entry.Resource == nil {
+
+			refURL := entry.FullUrl
+			if refURL != "" && entry.Search != nil && len(entry.Link) > 0 {
+				score := *entry.Search.Score
+				// Allow for possibility of true negatives being expressed in answer key
+				if score > 0 {
+					for _, link := range entry.Link {
+						if strings.EqualFold("related", link.Relation) {
+							linkedURL := link.Url
+							logger.Log.WithFields(logrus.Fields{
+								"answer":   i,
+								"full url": refURL,
+								"link url": linkedURL,
+								"search":   score}).Info("buildAnswerMap")
+							// look for existing entry
+							item := m[refURL]
+							if item == nil {
+								item = &groundTruth{}
+								item.linkedURLs = make([]string, 1)
```

Comment:
not a really bug deal, but in the future, I think we can shorten this with slice literals

```go
item.linkedURLs = []string{linkedURL}
item.numFound = []int{1}
```